### DISCUSSION TITLE
如何在protected_router.js里添加私有的路由

### DISCUSSION THREAD

**Discussion by shawn8888** [DIYgod/RSSHub#8853] [Q&A] [answered]
本人初学javascript，想试试自己写两个简单的路由练练手。
看了”参与我们”中的这段：
```

添加脚本路由
在 /lib/router.js (opens new window) 里添加路由
# 编写脚本
在 /lib/routes/ (opens new window) 中的路由对应路径下创建新的 js 脚本：

```
很详细！于是自己写了一个简单的路由，但因为只是自己测试使用，所以想放在protected_router.js中，这样不受docker更新的影响。但是找来找去，没找到关于protected_router的说明文档。
请问：
1. 如果使用官方的docker，怎样映射protected_router才能保留自己写的路由不会被更新覆盖？
2. protected_router.js应该怎样写？ 能否给个例子？router.js里有很多的例子可以照葫芦画瓢，但protected_router.js中找不到例子，无从下手，苦恼。

谢谢！

**Comment by syrinka** [comment, marked as answer]
```js
// 修改 protected_router.js
// 这里我把身份认证部分删掉了
const Router = require('@koa/router');
const router = new Router();

// 实际路由为 /protected/path/to/site
router.get('/path/to/site', require('./private/site'));

module.exports = router;
```

写法参照 https://github.com/DIYgod/RSSHub/blob/master/lib/router.js 就好

部署时加上两行挂载命令

```yaml
volumes:
- {修改后的 protected_router.js}:/app/lib/protected_router.js
- {私有路由放置文件夹}:/app/lib/private
```

**Reply by appleftg** [reply]
挂载的protected_router.js应该是protected-router.js，顺利找到本地自定义理由，非常感谢，不知官方的配置说明在哪，没有找到protected-router的介绍用法。

**Comment by gkeyes** [comment]
为什么我尝试了下
使用`router.get('/path/to/site', require('./private/site'));` ，无法访问路由，路由文件夹移动到v2后
使用`router.get('/path/to/site', require('./v2/site'));`，可以正常访问

**Reply by gkeyes** [reply]
应该是升级V2 和之前不一样了。。。。
替换下v2router.js 的内容，就可以将其他文件夹里的V2 私人路由挂载放进来了。
```

const path = require('path');
const requireAll = require('require-all');

const folders = ['private', 'v2'];
const routes = {};

folders.forEach(folder => {
    const dirname = path.join(__dirname, folder);

    // 遍历整个 routes 文件夹，收集模块路由 router.js
    const RouterPath = requireAll({
        dirname,
        filter: /router\.js$/,
    });

    // 将收集到的自定义模块路由进行合并
    for (const dir in RouterPath) {
        const project = RouterPath[dir]['router.js']; // Do not merge other file
        routes[`${dir}`] = project;
    }
});

module.exports = routes;

```

**Comment by libukai** [comment]
## 文件映射

以 Docker Compose 部署为例，使用私有路由需要映射路由空间和注册文件到容器中 `/app/lib` 下对应的路径。

```yaml title="docker-compose.yml"
volumes:
    - {/path/of/your/private}:/app/lib/private
    - {/path/of/your/protected-router.js}:/app/lib/protected-router.js
```

如果是在本地进行调试，则需要在 `/lib` 文件夹下新建 `/private` 文件夹，存放新的路由空间，并修改 `protected-router.js` 文件注册新的路由。

## 创建规则

因为私有路由不涉及到对外的分享，所以路由空间下不用创建 `maintainer.js`，也不用创建 `radar.js` 文件。此外，由于私有路由会单独通过 `protected-router.js` 文件单独注册，因此也不需要创建 `router.js` 文件。

所以，私有路由必须要创建的只有 RSS 源规则文件，这个规则的要求和公开路由的创建规则一致，详情请查看 [制作自己的 RssHub 路由](https://docs.rsshub.app/zh/joinus/new-rss/start-code)。

## 注册路由

通过修改 `protected-router.js` 文件，可以单独注册和管理私有路由。

注册私有路由的语法和公开路由一致，详情请查看 [router.js 配置说明](https://docs.rsshub.app/zh/joinus/new-rss/before-start#%E7%90%86%E8%A7%A3%E5%9F%BA%E7%A1%80%E7%9F%A5%E8%AF%86)

这里注意一下，注册私有路由的时候，需要在路由路径前加上 `/private` 字段，以指向正确的路由空间路径。

```javascript hl_lines="5"
const Router = require('@koa/router');
const router = new Router();

// NYTimes
router.get('/nytimes/:channel?', require('./private/nytimes/channel'));

module.exports = router;
```

为了保护私有路由，可以在 `protected-router.js` 文件中为路由添加 Auth 授权认证。

```javascript hl_lines="4 5 6"
const Router = require('@koa/router');
const router = new Router();

const auth = require('koa-basic-auth');
// 修改为自己的用户名和密码
const config = { name: 'rss', pass: 'hub' };
router.use('/(.*)', auth(config));

// NYTimes
router.get('/nytimes/:channel?', require('./private/nytimes/channel'));

module.exports = router;
```

## 访问路由

访问私有路由的方式和公开路由基本一致，但需要在路径中的路由空间前添加 `/protected` 字段。

```html
https://rsshub.example.com/protected/nytimes
```

如果添加了授权认证，则需要在链接中添加用户名和密码。

```html
https://rss:hub@rsshub.example.com/protected/nytimes
```

在浏览器中测试需要授权认证的私有路由时，经常会遇到浏览器缓存的问题导致失败，推荐使用在终端中使用 `curl` 进行测试。

```bash title="终端测试"
curl -u rss:hub https://rsshub.example.com/protected/nytimes
```

**Reply by libukai** [reply]
图文版本请查看 [爱可帮·AI CAN HELP](https://aikebang.net/pages/rsshub/rsshub%E7%A7%81%E6%9C%89%E8%B7%AF%E7%94%B1/)

**Comment by mpco** [comment]
这个方法已经失效了吧，现在找不到protected_router.js了
